e2e/read-initial-options/ts-esbuild-register-config/jest.config.ts

```diff
@@ -0,0 +1,11 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @jest-config-loader esbuild-register
+ */
+export default {
+  jestConfig: 'jest.config.ts',
```

Comment:
Add a type annotation? Currently this file is valid JS

Reply:
Makes sense, done in 66e69a1d5d59fdc3e32d0a148a23b2a5672283d4.

Reply:
I can’t find the exact failure, but this note could be helpful. Only one job in CI builds types, all other tests run without types being build. So if you import some type from `@jest/types`, type check will fail.

Integration tests for `ts-node` used to read Jest config with type checks are ignored in `jest.config.msj` and run through `jest.config.ts.mjs` (with types build): 

https://github.com/facebook/jest/blob/d2420aaf42055097dfe1e6a54bf3701214b06402/jest.config.mjs#L67

Reply:
Ah, interesting.

Should I add testcases to `e2e/__tests__/tsIntegration.test.ts` for the new functionality? The testcases that check for a TS type failure won't fail the same way under `esbuild-register` since it doesn't do type checking, so would need to do different error checking.

For reference:
```
● using esbuild-register › when `Config` type is imported from "@jest/types" › throws if type errors are encountered

    expect(received).toMatch(expected)

    Expected substring: "jest.config.ts(3,40): error TS2322: Type 'string' is not assignable to type 'number'."
    Received string:    "● Validation Error:·
      Option \"testTimeout\" must be of type:
        number
      but instead received:
        string·
      Example:
      {
        \"testTimeout\": 5000
      }·
      Configuration Documentation:
      https://jestjs.io/docs/configuration
    "

      77 |       const {stderr, exitCode} = runJest(DIR);
      78 |
    > 79 |       expect(stderr).toMatch(
         |                      ^
      80 |         "jest.config.ts(3,40): error TS2322: Type 'string' is not assignable to type 'number'.",
      81 |       );
      82 |       expect(exitCode).toBe(1);

      at Object.toMatch (e2e/__tests__/tsIntegration.test.ts:79:22)

  ● using esbuild-register › when `Config` type is imported from "@jest/types" › throws if syntax errors are encountered

    expect(received).toMatch(expected)

    Expected substring: "jest.config.ts(4,16): error TS2304: Cannot find name 'get'."
    Received string:    "Error: Jest: Failed to parse the TypeScript config file /Users/mpeveler/code/github/jest/e2e/ts-node-integration/jest.config.ts
      Error: Transform failed with 1 error:
    /Users/mpeveler/code/github/jest/e2e/ts-node-integration/jest.config.ts:4:19: ERROR: Expected \";\" but found \"config\"
        at readConfigFileAndSetRootDir (/Users/mpeveler/code/github/jest/packages/jest-config/build/readConfigFileAndSetRootDir.js:123:13)
        at async readInitialOptions (/Users/mpeveler/code/github/jest/packages/jest-config/build/index.js:396:13)
        at async readConfig (/Users/mpeveler/code/github/jest/packages/jest-config/build/index.js:147:48)
        at async readConfigs (/Users/mpeveler/code/github/jest/packages/jest-config/build/index.js:417:26)
        at async runCLI (/Users/mpeveler/code/github/jest/packages/jest-core/build/cli/index.js:144:59)
        at async Object.run (/Users/mpeveler/code/github/jest/packages/jest-cli/build/run.js:124:37)"

       98 |       const {stderr, exitCode} = runJest(DIR);
       99 |
    > 100 |       expect(stderr).toMatch(
          |                      ^
      101 |         "jest.config.ts(4,16): error TS2304: Cannot find name 'get'.",
      102 |       );
      103 |       expect(exitCode).toBe(1);

      at Object.toMatch (e2e/__tests__/tsIntegration.test.ts:100:22)

  ● using esbuild-register › when `Config` type is imported from "@jest/types" › throws if type errors are encountered when package.json#type=module

    expect(received).toMatch(expected)

    Expected substring: "jest.config.ts(3,42): error TS2322: Type 'string' is not assignable to type 'number'."
    Received string:    "● Validation Error:·
      Option \"testTimeout\" must be of type:
        number
      but instead received:
        string·
      Example:
      {
        \"testTimeout\": 5000
      }·
      Configuration Documentation:
      https://jestjs.io/docs/configuration
    "

      159 |       const {stderr, exitCode} = runJest(DIR);
      160 |
    > 161 |       expect(stderr).toMatch(
          |                      ^
      162 |         "jest.config.ts(3,42): error TS2322: Type 'string' is not assignable to type 'number'.",
      163 |       );
      164 |       expect(exitCode).toBe(1);

      at Object.toMatch (e2e/__tests__/tsIntegration.test.ts:161:22)

  ● using esbuild-register › when `Config` type is imported from "@jest/types" › throws if syntax errors are encountered when package.json#type=module

    expect(received).toMatch(expected)

    Expected substring: "jest.config.ts(4,16): error TS2304: Cannot find name 'get'."
    Received string:    "Error: Jest: Failed to parse the TypeScript config file /Users/mpeveler/code/github/jest/e2e/ts-node-integration/jest.config.ts
      Error: Transform failed with 1 error:
    /Users/mpeveler/code/github/jest/e2e/ts-node-integration/jest.config.ts:4:21: ERROR: Expected \";\" but found \"config\"
        at readConfigFileAndSetRootDir (/Users/mpeveler/code/github/jest/packages/jest-config/build/readConfigFileAndSetRootDir.js:123:13)
        at async readInitialOptions (/Users/mpeveler/code/github/jest/packages/jest-config/build/index.js:396:13)
        at async readConfig (/Users/mpeveler/code/github/jest/packages/jest-config/build/index.js:147:48)
        at async readConfigs (/Users/mpeveler/code/github/jest/packages/jest-config/build/index.js:417:26)
        at async runCLI (/Users/mpeveler/code/github/jest/packages/jest-core/build/cli/index.js:144:59)
        at async Object.run (/Users/mpeveler/code/github/jest/packages/jest-cli/build/run.js:124:37)"

      180 |       const {stderr, exitCode} = runJest(DIR);
      181 |
    > 182 |       expect(stderr).toMatch(
          |                      ^
      183 |         "jest.config.ts(4,16): error TS2304: Cannot find name 'get'.",
      184 |       );
      185 |       expect(exitCode).toBe(1);
```

Reply:
> Should I add testcases to `e2e/__tests__/tsIntegration.test.ts` for the new functionality?

yes please 👍